_posts/2023-01-16-native-adopts-adaptive-gc-policy.adoc

```diff
@@ -0,0 +1,149 @@
+2023-01-09 13:29:32,155 INFO  [io.quarkus] (main) quarkus-project 1.0.0-SNAPSHOT native (powered by Quarkus 2.15.2.Final) started in 0.017s. Listening on: http://0.0.0.0:8080
+...
+[Heap policy parameters:
+  YoungGenerationSize: 268435456
+      MaximumHeapSize: 27487790640
+      MinimumHeapSize: 536870912 <--
+     AlignedChunkSize: 1048576
+  LargeArrayThreshold: 131072]
+----
+
+We realized that this number is actually 512MB,
+which is the default minimum heap size GraalVM configures when the maximum heap size is anything above ~3GB of physical memory.
+
+The next question was,
+why is there a relationship between the minimum heap size and the memory consumption at which a full GC appears to happen?
+Looking at the output above,
+on our system the default maximum heap size is 25.6GB.
+GraalVM defaults the maximum heap size to 80% of the physical memory if no specific configuration is passed, and indeed 25.6GB is 80% of 32GB.
+It would seem odd to do a full GC when 512MB have been consumed,
+given that our system has given it a maximum heap size that is far bigger.
+The answer was found in the GC policy Quarkus was explicitly configuring.
+
+By default GraalVM configures a GC policy called "adaptive",
+but Quarkus was instead instructing GraalVM to use another GC policy called "by space and time".
+The full story on why Quarkus was using a different GC policy can be found
+https://github.com/quarkusio/quarkus/issues/28267[here],
+but to summarize,
+the decision was made in 2018, when "by space and time" appeared to generate less full GCs and offered considerably better throughput.
+
+The "by space and time" GC policy implemented a `shouldCollectCompletely` method that decided whether to do a complete (full) or incremental (minimal) collection.
+The relevant code of the "by space and time" GC policy is the following:
+
+[source,java]
+----
+return estimateUsedHeapAtNextIncrementalCollection().aboveThan(getMaximumHeapSize()) // (1)
+  || GCImpl.getChunkBytes().aboveThan(getMinimumHeapSize()) && enoughTimeSpentOnIncrementalGCs(); // (2)
+----
+
+One option `(1)` for doing a full GC would be when it estimates that the used heap will exceed maximum heap size,
+but that wasn’t our case.
+The other `(2)` would be if enough minimal collections had happened and the used heap was above the minimum heap size.
+This latter option was what was happening here.
+
+At this point we thought,
+do the assumptions made about the default GC policy still apply in 2022?
+So, we removed the GC policy configuration tweak,
+repeated the test and we observed the following memory consumption:
+
+image::adaptive-memory-consumption.png[]
+
+For the same workload the default GC policy, called "adaptive",
+consumed close to 50% less heap compared to the "by space and time" one.
+Note, however, that these graphs alone are not enough to make the switch since we could have a situation where adaptive is using less memory because the overall throughput is less.
```

Comment:
```suggestion
Note, however, that these graphs alone are not enough to make the switch since we could have a situation where "adaptive" is using less memory because the overall throughput is less.
```